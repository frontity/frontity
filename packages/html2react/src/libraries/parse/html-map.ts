const attributes: { [key: string]: string } = [
  "acceptCharset",
  "accessKey",
  "allowFullScreen",
  "allowTransparency",
  "autoComplete",
  "autoFocus",
  "autoPlay",
  "cellPadding",
  "cellSpacing",
  "charSet",
  "classID",
  "className",
  "colSpan",
  "contentEditable",
  "contextMenu",
  "crossOrigin",
  "dateTime",
  "encType",
  "formAction",
  "formEncType",
  "formMethod",
  "formNoValidate",
  "formTarget",
  "frameBorder",
  "hrefLang",
  "htmlFor",
  "httpEquiv",
  "inputMode",
  "keyParams",
  "keyType",
  "marginHeight",
  "marginWidth",
  "maxLength",
  "mediaGroup",
  "minLength",
  "noValidate",
  "radioGroup",
  "readOnly",
  "rowSpan",
  "spellCheck",
  "srcDoc",
  "srcLang",
  "srcSet",
  "tabIndex",
  "useMap"
].reduce((map, value) => {
  map[value.toLowerCase()] = value;
  return map;
}, {});

export default attributes;
