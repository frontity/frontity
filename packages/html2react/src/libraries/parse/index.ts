import { parse as himalaya } from "himalaya";
import {
  HimalayaElement,
  HimalayaText,
  HimalayaComment,
  HimalayaNode
} from "../../../himalaya/types";
import he from "he";
import htmlMap from "./html-map";
import svgMap from "./svg-map";

// Adapts the Himalaya AST Specification v1
// (see https://github.com/andrejewski/himalaya/blob/v1.0.1/text/ast-spec-v1.md)
// to our format, with the following structure:
//
// {
//   type: 'element',
//   component: 'p',
//   props: {},
//   children: [],
//   parent: Element,
// }

interface Element {
  type: "element";
  component: string;
  props: { [key: string]: string | boolean };
  children?: Node[];
  parent?: Element;
}

interface Text {
  type: "text";
  content: string;
  parent?: Element;
}

interface Comment {
  type: "comment";
  content: string;
  parent?: Element;
}

type Node = Element | Text | Comment;

// Map of lowercased HTML and SVG attributes to get their camelCase version.
const attrMap = { ...htmlMap, ...svgMap };

function adaptNode(element: HimalayaNode, parent?: Element): Node {
  let node: Node;

  if (element.type === "element") {
    node = {
      type: element.type,
      component: element.tagName,
      props: element.attributes.reduce(
        (props: Element["props"], { key, value }) => {
          if (key === "class") {
            props.className = value;
          } else if (key === "style") {
            props["data-style"] = value;
          } else if (!/^on/.test(key)) {
            const camelCaseKey = attrMap[key.toLowerCase()];
            // Map keys with no value to `true` booleans.
            props[camelCaseKey || key] = value === null ? true : value;
          }

          return props;
        },
        {}
      )
    };

    node.children = element.children.reduce(
      (final: Node[], child: any): Node[] => {
        const childAdapted = adaptNode(child, node as Element);
        // ignore null children
        if (childAdapted) final.push(childAdapted);
        return final;
      },
      []
    );
  }

  if (element.type === "text") {
    const content = he.decode(element.content);

    if (content.trim().length) {
      node = {
        type: element.type,
        content: element.content
      };
    } else return null;
  }

  if (element.type === "comment") {
    const content = he.decode(element.content);

    if (content.trim().length) {
      node = {
        type: element.type,
        content: element.content
      };
    } else return null;
  }

  if (parent) node.parent = parent;

  return node;
}

// Parse HTML code using Himalaya's parse function first, and then
// adapting each node to our format.
const parse = (html: string): Node[] =>
  himalaya(html).reduce((htmlTree: Node[], element) => {
    const adapted = adaptNode(element);
    if (adapted) htmlTree.push(adapted);
    return htmlTree;
  }, []);

export default parse;
