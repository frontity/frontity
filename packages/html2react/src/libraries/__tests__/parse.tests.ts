import parse from "../parse";

describe("parse", () => {
  const html = `
    <!-- This is a comment -->
    <p id="paragraph">This is an html <span>example</span> for testing purposes.</p>

    <!-- [class] attribute -->
    <div class="hello">
      <img src="http://example.com/img.jpg">
    </div>
    
    <!-- [autoplay] standalone attribute ([key/value] should be [autoPlay/true] -->
    <audio autoplay>
      <source src="example.ogg" type="audio/ogg">
    </div>
  `;

  test("should return the right HTML tree", () => {
    const result = parse(html);
    expect(result).toMatchSnapshot();
  });
});
