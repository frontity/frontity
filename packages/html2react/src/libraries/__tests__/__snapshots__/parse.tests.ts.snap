// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should return the right HTML tree 1`] = `
Array [
  Object {
    "content": " This is a comment ",
    "type": "comment",
  },
  Object {
    "children": Array [
      Object {
        "content": "This is an html ",
        "parent": [Circular],
        "type": "text",
      },
      Object {
        "children": Array [
          Object {
            "content": "example",
            "parent": [Circular],
            "type": "text",
          },
        ],
        "component": "span",
        "parent": [Circular],
        "props": Object {},
        "type": "element",
      },
      Object {
        "content": " for testing purposes.",
        "parent": [Circular],
        "type": "text",
      },
    ],
    "component": "p",
    "props": Object {
      "id": "paragraph",
    },
    "type": "element",
  },
  Object {
    "content": " [class] attribute ",
    "type": "comment",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "img",
        "parent": [Circular],
        "props": Object {
          "src": "http://example.com/img.jpg",
        },
        "type": "element",
      },
    ],
    "component": "div",
    "props": Object {
      "className": "hello",
    },
    "type": "element",
  },
  Object {
    "content": " [autoplay] standalone attribute ([key/value] should be [autoPlay/true] ",
    "type": "comment",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "source",
        "parent": [Circular],
        "props": Object {
          "src": "example.ogg",
          "type": "audio/ogg",
        },
        "type": "element",
      },
    ],
    "component": "audio",
    "props": Object {
      "autoPlay": true,
    },
    "type": "element",
  },
]
`;
